### ISSUE TITLE
Expose a virtual BMC service per virtual machine, for IPMI and/or Redfish actions

### ISSUE THREAD

**Issue by JM1** [kubevirt/kubevirt#9108]
**Is your feature request related to a problem? Please describe**:
OpenShift and OpenStack numerous installers allow to deploy clusters to bare-metal servers and clouds. For development, debugging, testing etc. it helps to run these installers in local virtual environments. For example, a developer can deploy OpenShift and OpenStack on/to libvirt domains or qemu virtual machines running on her/his system locally. This requires setting up an BMC emulator such as [virtualbmc](https://docs.openstack.org/virtualbmc/latest/) (for IPMI) or [sushy-emulator](https://docs.openstack.org/sushy-tools/latest/user/dynamic-emulator.html) (for Redfish) which handles tasks such as virtual machine reboots and virtual media mounts.

Now, the overall goal is to Kubernetes-ize (containerize with K8s) these local virtualization-based deployments: Kubernetes+KubeVirt instead of libvirtd+virtualbmc/sushy-emulator. For example, setting up a [k3s instance in Docker](https://docs.k3s.io/advanced#running-k3s-in-docker) with [KubeVirt](https://charlottemach.com/2020-11-03-windows-kubevirt-k3s.html) and launching KubeVirt VMs is straight forward. What is missing though is a BMC emulator. Both virtualbmc and sushy-emulator require access to `libvirtd`, but iiuc KubeVirt's own `virt-launcher` is using `virtqemud` inside the [VMI/virt-launcher pod](https://github.com/kubevirt/kubevirt/blob/main/docs/architecture.md) only, so virtualbmc and sushy-emulator cannot be used.

When KubeVirt would provide a BMC emulator for each virtual machine, then we could deploy production-like (*) OpenShift or OpenStack clusters in KubeVirt VMs using their native bare-metal installers. Paired with k3s and Docker/Podman this would allow us to spin up and destroy a virtualized OpenShift or OpenStack cluster in a single container easily.

(*) Multi node, same cluster features, same installer and using IPMI/Redfish.

**Describe the solution you'd like**:
A BMC emulator, speaking Redfish and/or IPMI protocols, should be provided by KubeVirt. KubeVirt should exposes this BMC emulator as a Kubernetes services to perform actions such as rebooting a virtual machine, changing its boot devices and mounting virtual media.

**Describe alternatives you've considered**:

* libvirt(d) based virtual machines with BMC emulators [virtualbmc](https://docs.openstack.org/virtualbmc/latest/) (for IPMI) or [sushy-emulator](https://docs.openstack.org/sushy-tools/latest/user/dynamic-emulator.html) (for Redfish).
* [kcli](https://kcli.readthedocs.io/en/latest/) allows to deploy various Kubernetes flavors in libvirtd based virtual machines. 
  - Basically a wrapper for libvirt based approach.
* [CRC](https://crc.dev/crc/) allows to deploy a minimal OpenShift locally with Podman.
  - Subset of OpenShift features only.
  - No IPMI/Redfish installer.
* [kind](https://kind.sigs.k8s.io/) allows to deploy a Kubernetes cluster locally in Docker containers.
  - No OpenShift nor OpenStack support.
  - No IPMI/Redfish installer.
* [OpenShift supports installation to public clouds such as Amazon, Azure, Google or OpenStack clouds](https://docs.openshift.com/container-platform/4.12/installing/installing-preparing.html).
  - Not a local deployment on developer machine.

**Comment by alancaldelas** [comment]
How would this work? I know nothing.

**Comment by fabiand** [comment]
I think it's a great idea to have a BMC for KubeVirt.

@colonwq iirc you worked on something like this on https://github.com/colonwq/virtualbmc ?

By today we'd probably want to have a redfish implementation.

@jm1 do you have capcity to contribute such a feature?

**Comment by JM1** [comment]
/remove-lifecycle stale

The OpenStack Ironic folks wrote a Redfish Emulator as part of [sushy-tools](https://github.com/openstack/sushy-tools). It offers a Redfish API and internally translates it to libvirtd. (OpenStack Ironic is also used in OpenShift, e.g. for IPI).

Some time ago I had a brief look of how this could be implemented in kubevirt (not so easy without access to libvirtd), but had to stop due to lack of time.

**Comment by JM1** [comment]
[virtualbmc](https://docs.openstack.org/virtualbmc/latest/) is similar to sushy-tools in the sense that it translates IPMI API to libvirtd calls.

**Comment by alancaldelas** [comment]
https://github.com/starbops/kubebmc

^ this seems like what we are trying to do right?

**Comment by JM1** [comment]
@alancaldelas Thank you for pointing us to [KubeBMC](https://github.com/starbops/kubebmc) 🤩 Indeed it answers this issue partially!

It uses [VMware's Go IPMI library](github.com/vmware/goipmi) and its simulator in order to [provide a subset of the IPMI API](https://github.com/starbops/kubebmc/blob/7a122d0461c0604716a707609816a8b0c660c274/pkg/kbmc/kbmc.go#L44), mainly powering on/off a VM, querying the power state and setting the boot device to disk or pxe. For easier usage and distribution it would be great if it would be incorporated into KubeVirt.

For virtual media support, a Redfish API would still be necessary though.

**Comment by fabiand** [comment]
I'd at least offer a KubeVirt T-Shirt if somebody is helping this project to come up with a RedFish implementation/backend for KubeVirt.

**Comment by xpivarc** [comment]
A proposal was merged, have the work on this started?

**Comment by alancaldelas** [comment]
I am interested in picking this up, I am super new though what steps are needed for me to get started?

**Comment by lentzi90** [comment]
From what I understand we are now almost finished with [phase 1](https://github.com/kubevirt/community/blob/main/design-proposals/kubevirtbmc.md#phase-1-transfer-the-kubebmc-project-to-the-kubevirt-organization).

- The CRDs have been [changed to conform with KubeVirt conventions](https://github.com/starbops/kubevirtbmc/issues/5)
- A [ton of tests has been added](https://github.com/starbops/kubevirtbmc/issues/6)

If the KubeVirt community agrees that this is enough, I think we are ready to transfer the repo to the kubevirt org.

After that, it is time for [phase 2](https://github.com/starbops/kubevirtbmc/issues/6). Basically adapting/implementing the proposed design.

**Comment by starbops** [comment]
Happy New Year, everyone! As [phase 1](https://github.com/starbops/kubevirtbmc/milestone/1) for [KubeVirtBMC](https://github.com/starbops/kubevirtbmc) is soon to be completed, I'd like to know the next step to start the project transfer. Is there any checkpoint that I should pass first? Thank you all!

**Comment by JM1** [comment]
FYI A new aspirant is [kubevirt-redfish](https://github.com/v1k0d3n/kubevirt-redfish). Thanks, @0xFelix, for bringing this to my attention.

**Comment by lentzi90** [comment]
I would still love to see this move forward. Kubevirtbmc (https://github.com/kubevirtbmc/kubevirtbmc) seems to be well maintained and in good shape. Is there anything I could help with to revive this issue?

**Comment by 0xFelix** [comment]
There is another project which got some attention in the recent months and is on track to be moved under the KubeVirt org, see

https://github.com/kubevirt/enhancements/pull/177
https://github.com/v1k0d3n/kubevirt-redfish/

**Comment by starbops** [comment]
@lentzi90 Thank you for bringing this up again. Your feedback and support over the past couple of years have meant a lot.

As @0xFelix noted, the KubeVirt community is moving forward with a different project. KubeVirtBMC will continue to evolve as an independent project at https://github.com/kubevirtbmc/kubevirtbmc. We see real value in keeping it going, and there are users already relying on it.

If you'd still like to help, contributions, issues, and ideas are very welcome over there. Happy to chat further in that repo 🙂

**Comment by 0xFelix** [comment]
See https://github.com/kubevirt/redfish-controller
